Rejections handling

juagargi opened the pull request:
I forgot to say:
This PR handles the case when one or more APs are unable to fulfill the requested change (or deletion, or...). 
In this case, the administrator(s) get notified, and the pending change is deleted.

The corresponding changes in the update gen script are done in PR https://github.com/netsec-ethz/scion-box/pull/36. Please check said PR and tell me I'm not crazy.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion-coord/224)
<!-- Reviewable:end -->

chaehni commented:
Reviewed 5 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[controllers/api/scion_lab_ap.go, line 187 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LCELHup1qWXQJhwRTEL:-LCELHup1qWXQJhwRTEM:bc6onr7) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L187)):*
> ```Go
> 	}
> 	body := string(bodyBytes)
> 	fmt.Println("Body:", body)
> ```

Is this for debugging or should it go into the log? 

---

*[email/email.go, line 53 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LC47EeS5kV3o6Ht79g6:-LC47EeT7ODCSqfPm2Y1:b2jncn6) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/email/email.go#L53)):*
> ```Go
> // their userEmail by filling in the specified template with the given subject
> // and information in the data object
> func Construct(emailTemplate, subject string, data interface{}, tag, userEmail string) (*Email, error) {
> ```

Does this function need to be public? It's only used in the email package.

---

*[email/email.go, line 110 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LCE4e-E0m3SM8aooDQE:-LCE4e-FDaXaaqDjIDVp:b4tgfkf) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/email/email.go#L110)):*
> ```Go
> 	}
> 
> 	if alsoToAdmins {
> ```

This only allows sending emails to admins when sending an email to a user at the same time. In the future, it might be useful to send emails to admins only. Do we have another mechanism for that?

---

*[email/email.go, line 123 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LCE5a4v5Cao07g0mN7I:-LCE5a4v5Cao07g0mN7J:b-wxobm9) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/email/email.go#L123)):*
> ```Go
> 
> // ConstructAndSend builds and sends an email only to the specified address; it relies on ConstructAndSendEmail
> func ConstructAndSend(emailTemplate string, subject string, data interface{},
> ```

I think we should only have one public ConstructAndSend...() function or give them a more descriptive name to make it clear which one forwards the email to the admins.

---


*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224)*
<!-- Sent from Reviewable.io -->

chaehni commented:
I had a look at the changes. I left some minor comments.

---

Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-:-LCTtcIj7XumH453jN5E:b-ieqnj9)*
<!-- Sent from Reviewable.io -->

mlegner commented:
I made a pass to and added some comments below.

---

Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[conf/development.conf.default, line 51 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEK4a56BcFCY3MlZ_v:-LEEK4a666DHJQTIAN0i:bzd4btd) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/conf/development.conf.default#L51)):*
> ```
> email.pm_server_token = ""
> email.pm_account_token = ""
> email.admin_emails = "admin1@scionlab.org; admin2@scionlab.org"
> ```

Maybe add some information what these email addresses are used for.
Also, if you remove the spaces here, we would not need to do this in the go code.

---

*[config/config.go, line 114 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEKEK56G0MWrN7_zQC:-LEEKEK56G0MWrN7_zQD:btzvhey) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/config/config.go#L114)):*
> ```Go
> 		SIGNING_ASES[ki] = vi
> 	}
> 	// validate email addresses?
> ```

I'm not sure that "validate" is the right expression here, all you are doing is removing spaces. 
Anyway, this would not be necessary if we entered the email addresses without spaces to begin with.
Finally, we could modify the slice in place instead of creating a new one.

---

*[controllers/api/scion_lab_ap.go, line 158 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEM_qs6oBmmlnKo6_8:-LEEM_qs6oBmmlnKo6_9:b7hy4wl) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L158)):*
> ```Go
> }
> 
> type attachedAsAck struct {
> ```

Should this mean "attached as ack" or "attached AS ack"? Anyway, I would prefer a more descriptive name.

---

*[controllers/api/scion_lab_ap.go, line 174 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEMH4cAxnVJ_GwAj-m:-LEEMH4d59oPR8eTz1-5:bcnx96f) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L174)):*
> ```Go
> //        {"Created":[],
> //         "Removed":[],
> //         "Updated":[{"IA": "1-1020", "success": true}, {"IA": "1-1020", "success": false}]
> ```

This PR should mention somehow, where the respective changes are implemented on the AP side.

---

*[controllers/api/scion_lab_ap.go, line 246 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEENP0q0Lp8euikJznN:-LEENP0q0Lp8euikJznO:b-1ntn4u) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L246)):*
> ```Go
> }
> 
> // 3-1034,[141.44.25.146]:30100
> ```

What is the purpose of this line?

---

*[controllers/api/scion_lab_ap.go, line 342 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEOh-D7qXpUgDiKdZN:-LEEOh-D7qXpUgDiKdZO:b-4xwjkj) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L342)):*
> ```Go
> }
> 
> func (s *SCIONLabASController) processRejectedUpdatesFromAP(rejections []rejectedAS) []string {
> ```

Maybe add a short comment describing what this function is supposed to do.

---

*[controllers/api/scion_lab_ap.go, line 345 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEENyS4AYJpyHLazlAb:-LEENyS4AYJpyHLazlAc:behwuf) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L345)):*
> ```Go
> 	failedNotifications := []string{}
> 	// for each rejected AS, send an email to the admin and user
> 	for _, ras := range rejections {
> ```

Does `ras` stand for `rejectedAS`? Maybe use `rAS` or `rejAS`.

---

*[controllers/api/scion_lab_ap.go, line 354 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEOWK47-F0Hb4jYqnM:-LEEOWK47-F0Hb4jYqnN:bngvh05) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L354)):*
> ```Go
> 		ap, err := models.FindSCIONLabASByIAString(ras.AP)
> 		if err != nil {
> 			log.Printf("Error finding SCIONLabAS %v: %v", ras.IA, err)
> ```

Should this be `ras.AP`?

---

*[controllers/api/scion_lab_ap.go, line 388 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEdkkcAqU3-rokamo_:-LEEdkkcAqU3-rokamoa:bd4k569) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L388)):*
> ```Go
> 		// fix the status of the AS entry, if needed:
> 		switch as.Status {
> 		default: // nothing to do
> ```

Could we not simply continue in the default case similar to above?

---

*[controllers/api/scion_lab_ap.go, line 426 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEQaC92JJzfREeOgh8:-LEEQaCA4OFMtL2CDx6D:b41aufb) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L426)):*
> ```Go
> 
> 	var message string
> 	subject := "[SCIONLab] Failure, "
> ```

Why "Failure" in the confirmation email?

---

*[controllers/api/scion_lab_ap.go, line 465 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEQDAo9e_op50H3lSQ:-LEEQDAo9e_op50H3lSR:b-1qo3r9) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L465)):*
> ```Go
> 
> 	subject := fmt.Sprintf("[SCIONLab] Could not complete request for %s", userIA)
> 	message := fmt.Sprintf(`The request for your AS %s to be (%s) from the attachment point %s could not be completed. You need to check the configuration for your AS in the Coordinator. Ensure you select the option to Update and Download the ScionLab Configuration after reviewing it.
> ```

Instead of hardcoding that rather long message here, I would suggest creating a new template similar to "as_status.html".

---

*[email/email.go, line 97 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEER6IU0aw7tIDuuk0I:-LEER6IU0aw7tIDuuk0J:b-onvvme) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/email/email.go#L97)):*
> ```Go
> 
> // ConstructAndSendEmail builds and then sends an email. It relies on Construct and Send
> func ConstructAndSendEmail(emailTemplate string, subject string, data interface{},
> ```

Why do we separate this into two functions `ConstructAndSend` and `ConstructAndSendEmail`?

---


*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-:-LEEe0KiD9igj52tHwtM:b-dyyjbq)*
<!-- Sent from Reviewable.io -->

juagargi commented:
Thanks guys

---

Reviewed 1 of 6 files at r1.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[conf/development.conf.default, line 51 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEK4a56BcFCY3MlZ_v:-LERKtnu5SfrDO3Het7M:b-xdztqf) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/conf/development.conf.default#L51)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Maybe add some information what these email addresses are used for.
Also, if you remove the spaces here, we would not need to do this in the go code.
</blockquote></details>

Done.
Also added a comment in the description of the PR, because I have forgotten again to say what it does.

---

*[config/config.go, line 114 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEKEK56G0MWrN7_zQC:-LERMjiu4mHKxsGxAFJk:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/config/config.go#L114)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

I'm not sure that "validate" is the right expression here, all you are doing is removing spaces. 
Anyway, this would not be necessary if we entered the email addresses without spaces to begin with.
Finally, we could modify the slice in place instead of creating a new one.
</blockquote></details>

Done.

---

*[controllers/api/scion_lab_ap.go, line 158 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEM_qs6oBmmlnKo6_8:-LERNKZWD9ITUlTn3GPT:bctrbuo) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L158)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Should this mean "attached as ack" or "attached AS ack"? Anyway, I would prefer a more descriptive name.
</blockquote></details>

Called "attachedASAckMessage" now

---

*[controllers/api/scion_lab_ap.go, line 174 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEMH4cAxnVJ_GwAj-m:-LERO8p26839zYNOdNML:b-owj5hf) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L174)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

This PR should mention somehow, where the respective changes are implemented on the AP side.
</blockquote></details>

Updated description for this PR (visible on the "conversation" tab of the PR). Summary: PR #36 of scion-box should have the other side of the changes. Please check for confirmation.

---

*[controllers/api/scion_lab_ap.go, line 187 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LCELHup1qWXQJhwRTEL:-LEROReg-rBsNU2VXG6Z:b-hvw0c0) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L187)):*
<details><summary><i>Previously, chaehni (chaehni) wrote…</i></summary><blockquote>

Is this for debugging or should it go into the log? 
</blockquote></details>

You didn't see anything there :laughing:  . It's gone now.

---

*[controllers/api/scion_lab_ap.go, line 246 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEENP0q0Lp8euikJznN:-LERQA_A4RO-9SqToJ9j:b3unpfj) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L246)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

What is the purpose of this line?
</blockquote></details>

Is a vital piece of the application
https://youtu.be/eFAvOcuJyHY?t=1m36s
:laughing: 
It's gone now.

---

*[controllers/api/scion_lab_ap.go, line 342 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEOh-D7qXpUgDiKdZN:-LERRRsE00uOfBbEO2Z2:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L342)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Maybe add a short comment describing what this function is supposed to do.
</blockquote></details>

Done.

---

*[controllers/api/scion_lab_ap.go, line 345 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEENyS4AYJpyHLazlAb:-LERRX_1C5qgmR7gKpTn:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L345)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Does `ras` stand for `rejectedAS`? Maybe use `rAS` or `rejAS`.
</blockquote></details>

Done.

---

*[controllers/api/scion_lab_ap.go, line 354 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEOWK47-F0Hb4jYqnM:-LERRlIk4oXEDoXxZxfs:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L354)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Should this be `ras.AP`?
</blockquote></details>

Done.

---

*[controllers/api/scion_lab_ap.go, line 388 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEdkkcAqU3-rokamo_:-LERSK6n9JgyzaEPWbaK:b-qhry5e) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L388)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Could we not simply continue in the default case similar to above?
</blockquote></details>

the intention of the code is to have a faster condition than a
`if as.Status == models.UPDATE || as.Status == models.REMOVE`

The default is mandatory, but must do nothing. This seems to be non obvious, so I have replaced it with the equivalent if... mentioned above, and deleted the `switch` completely.

---

*[controllers/api/scion_lab_ap.go, line 426 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEQaC92JJzfREeOgh8:-LERT5FcAMKgXL088uyb:b-886gsx) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L426)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Why "Failure" in the confirmation email?
</blockquote></details>

Fixed

---

*[controllers/api/scion_lab_ap.go, line 465 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEQDAo9e_op50H3lSQ:-LERTeJq47Q_KzUfoTC5:b-uxnvnr) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L465)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Instead of hardcoding that rather long message here, I would suggest creating a new template similar to "as_status.html".
</blockquote></details>

We are using as_status.html already. 

---

*[email/email.go, line 53 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LC47EeS5kV3o6Ht79g6:-LERU8LWAc74O3GvqZ49:b-tk8gyx) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/email/email.go#L53)):*
<details><summary><i>Previously, chaehni (chaehni) wrote…</i></summary><blockquote>

Does this function need to be public? It's only used in the email package.
</blockquote></details>

non public now

---

*[email/email.go, line 97 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEER6IU0aw7tIDuuk0I:-LERUtQW1F71TgJm1sLk:b-jj19xd) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/email/email.go#L97)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Why do we separate this into two functions `ConstructAndSend` and `ConstructAndSendEmail`?
</blockquote></details>

One would be the signature for the default parameter. I don't know if this is a common idiom in go, so after this comment I have removed it.

---

*[email/email.go, line 110 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LCE4e-E0m3SM8aooDQE:-LERVOnE5ELVCK10qGh7:b-j7184l) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/email/email.go#L110)):*
<details><summary><i>Previously, chaehni (chaehni) wrote…</i></summary><blockquote>

This only allows sending emails to admins when sending an email to a user at the same time. In the future, it might be useful to send emails to admins only. Do we have another mechanism for that?
</blockquote></details>

We could use ConstructAndSend specifying only the admin addresses, but I just added now
ConstructAndSendEmailToAdmins

---

*[email/email.go, line 123 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LCE5a4v5Cao07g0mN7I:-LERXT7g1FIuiA2EuOVP:b-8in3nx) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/email/email.go#L123)):*
<details><summary><i>Previously, chaehni (chaehni) wrote…</i></summary><blockquote>

I think we should only have one public ConstructAndSend...() function or give them a more descriptive name to make it clear which one forwards the email to the admins.
</blockquote></details>

Yup, addressed from a previous comment.

---


*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-:-LERJwim6cnbE8Ul7V3_:b-y4fmbq)*
<!-- Sent from Reviewable.io -->

juagargi commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224)*
<!-- Sent from Reviewable.io -->

mlegner commented:
I made another pass, please see the two remaining comments below.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[controllers/api/scion_lab_ap.go, line 174 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEMH4cAxnVJ_GwAj-m:-LETIXiK0roOn4_zKCpa:b-2xrgfl) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L174)):*
<details><summary><i>Previously, juagargi (Juan A. García Pardo Giménez de los Galanes) wrote…</i></summary><blockquote>

Updated description for this PR (visible on the "conversation" tab of the PR). Summary: PR #36 of scion-box should have the other side of the changes. Please check for confirmation.
</blockquote></details>

OK. It is PR https://github.com/netsec-ethz/scion-box/pull/36.

---

*[controllers/api/scion_lab_ap.go, line 388 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEdkkcAqU3-rokamo_:-LETJHAtAd_ua98YvbiN:bnl04sr) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L388)):*
<details><summary><i>Previously, juagargi (Juan A. García Pardo Giménez de los Galanes) wrote…</i></summary><blockquote>

the intention of the code is to have a faster condition than a
`if as.Status == models.UPDATE || as.Status == models.REMOVE`

The default is mandatory, but must do nothing. This seems to be non obvious, so I have replaced it with the equivalent if... mentioned above, and deleted the `switch` completely.
</blockquote></details>

OK, that works too. I originally thought about a construction similar to line 372.

---

*[controllers/api/scion_lab_ap.go, line 465 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEQDAo9e_op50H3lSQ:-LETJyiiDgyv-FUvveCQ:b-bq47s8) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L465)):*
<details><summary><i>Previously, juagargi (Juan A. García Pardo Giménez de los Galanes) wrote…</i></summary><blockquote>

We are using as_status.html already. 
</blockquote></details>

Yes, I know. But I would find it a lot nice to create a new template "as_failure.html" that will be used in this case.

---

*[controllers/api/scion_lab_ap.go, line 407 at r2](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LETJka9CgvwzllfNrgX:-LETJka9CgvwzllfNrgY:b-ofsek) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d2e4e7a2955a44d97c463b893217dfff532f1194/controllers/api/scion_lab_ap.go#L407)):*
> ```Go
> 			}
> 		}
> 		if as.Update() != nil {
> ```

Where do we modify the AS? Do we even need to update it?

---


*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-:-LETKC_f-Ufk3Mj2mN8x:b73xgwa)*
<!-- Sent from Reviewable.io -->

chaehni commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
My comments are addressed.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-:-LETty7P83U3_VJpyn04:b-7tqr0o)*
<!-- Sent from Reviewable.io -->

juagargi commented:
Pushing more changes.

---

Review status: 7 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[controllers/api/scion_lab_ap.go, line 388 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEdkkcAqU3-rokamo_:-LEU1oDr0b66wQ5tiIJN:bolm0ae) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L388)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

OK, that works too. I originally thought about a construction similar to line 372.
</blockquote></details>

I have also removed that switch in 372, I think now it's more readable.

---

*[controllers/api/scion_lab_ap.go, line 465 at r1](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEEQDAo9e_op50H3lSQ:-LEU40Ld5nGz4BhBVqhz:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d1095106f69072001c53dd7c0700a63d18d6049e/controllers/api/scion_lab_ap.go#L465)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Yes, I know. But I would find it a lot nice to create a new template "as_failure.html" that will be used in this case.
</blockquote></details>

Done.

---

*[controllers/api/scion_lab_ap.go, line 407 at r2](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LETJka9CgvwzllfNrgX:-LEU06qxECq5S-WRcDc-:b-sym4zc) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d2e4e7a2955a44d97c463b893217dfff532f1194/controllers/api/scion_lab_ap.go#L407)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Where do we modify the AS? Do we even need to update it?
</blockquote></details>

Yes, we need to update the entry. From line 370 until end of the function what we do is find the failed operation, delete it, and change the status of that as entry, in line ~ 394

---


*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-:-LEU43BC4e4OUlcobChJ:bxyoluh)*
<!-- Sent from Reviewable.io -->

mlegner commented:
Added a few more comments below. :smiley: 

---

Review status: 7 of 9 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[controllers/api/scion_lab_ap.go, line 407 at r2](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LETJka9CgvwzllfNrgX:-LEULEsL1NE1jg3doAFR:bvwe60i) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d2e4e7a2955a44d97c463b893217dfff532f1194/controllers/api/scion_lab_ap.go#L407)):*
<details><summary><i>Previously, juagargi (Juan A. García Pardo Giménez de los Galanes) wrote…</i></summary><blockquote>

Yes, we need to update the entry. From line 370 until end of the function what we do is find the failed operation, delete it, and change the status of that as entry, in line ~ 394
</blockquote></details>

I'm not convinced. We only modify the AS status inside the if statement that starts at line 384. And in this if statement, we have another call to `as.Update()` in line 399.

---

*[controllers/api/scion_lab_ap.go, line 188 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEUMmIF5cLItdm_0PO3:-LEUMmIF5cLItdm_0PO4:b-4rsiv6) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/controllers/api/scion_lab_ap.go#L188)):*
> ```Go
> 	body := string(bodyBytes)
> 
> 	var UpdateLists map[string]map[string][]attachedASAckMessage
> ```

rename to `updateLists`

---

*[email/templates/as_failure.html, line 4 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEULvQV4LCACtYqvhvH:-LEULvQV4LCACtYqvhvI:b1yaah5) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/email/templates/as_failure.html#L4)):*
> ```HTML
> 
> 
> {{.Message}}
> ```

The message is never set.

---

*[email/templates/as_failure.html, line 5 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEUMtUDBjlAg1MIhyfn:-LEUMtUDBjlAg1MIhyfo:bn5nvfo) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/email/templates/as_failure.html#L5)):*
> ```HTML
> 
> {{.Message}}
> The request for your AS {{.AS}} to be {{.Operation}} from the attachment point {{.AP}} could not be completed. 
> ```

"created from the AP" or "updated from the AP" doesn't make so much sense. Maybe we can use "at" instead of "from" or reformulate the phrase.

---

*[email/templates/as_failure.html, line 6 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEUN5Az4SkRrZsf7lp4:-LEUN5Az4SkRrZsf7lp5:bcd0bgc) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/email/templates/as_failure.html#L6)):*
> ```HTML
> {{.Message}}
> The request for your AS {{.AS}} to be {{.Operation}} from the attachment point {{.AP}} could not be completed. 
> You probably want to check the configuration for your AS in the Coordinator. Ensure you select the option to Update and Download the ScionLab Configuration after reviewing it.
> ```

SCIONLab configuration; why capitalize "Update" and "Download"?

---

*[email/templates/as_failure.html, line 7 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEUN7Ta8yC-UjF60_Eh:-LEUN7Ta8yC-UjF60_Ei:b-b2nzqs) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/email/templates/as_failure.html#L7)):*
> ```HTML
> The request for your AS {{.AS}} to be {{.Operation}} from the attachment point {{.AP}} could not be completed. 
> You probably want to check the configuration for your AS in the Coordinator. Ensure you select the option to Update and Download the ScionLab Configuration after reviewing it.
> If the problem persists, please contact the ScionLab Administrators.
> ```

SCIONLab administrators

---


*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-:-LEUNQvf6aA76BHebQ_5:b-fnhsv9)*
<!-- Sent from Reviewable.io -->

juagargi commented:
Reviewed 2 of 2 files at r3.
Review status: 7 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[controllers/api/scion_lab_ap.go, line 407 at r2](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LETJka9CgvwzllfNrgX:-LEUXDl95NaGCDT8BKYq:bss4a1y) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/d2e4e7a2955a44d97c463b893217dfff532f1194/controllers/api/scion_lab_ap.go#L407)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

I'm not convinced. We only modify the AS status inside the if statement that starts at line 384. And in this if statement, we have another call to `as.Update()` in line 399.
</blockquote></details>

Removed!

---

*[controllers/api/scion_lab_ap.go, line 188 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEUMmIF5cLItdm_0PO3:-LEUWT9lAAUpm-dFMEL4:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/controllers/api/scion_lab_ap.go#L188)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

rename to `updateLists`
</blockquote></details>

Done.

---

*[email/templates/as_failure.html, line 4 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEULvQV4LCACtYqvhvH:-LEUVA0XCMEeTBC6dMEa:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/email/templates/as_failure.html#L4)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

The message is never set.
</blockquote></details>

Done.

---

*[email/templates/as_failure.html, line 5 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEUMtUDBjlAg1MIhyfn:-LEUVLQU8Vr6-TYbs6Z6:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/email/templates/as_failure.html#L5)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

"created from the AP" or "updated from the AP" doesn't make so much sense. Maybe we can use "at" instead of "from" or reformulate the phrase.
</blockquote></details>

Done.

---

*[email/templates/as_failure.html, line 6 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEUN5Az4SkRrZsf7lp4:-LEUVp3gEjFXDBu5wi3L:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/email/templates/as_failure.html#L6)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

SCIONLab configuration; why capitalize "Update" and "Download"?
</blockquote></details>

Done.

---

*[email/templates/as_failure.html, line 7 at r3](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-LEUN7Ta8yC-UjF60_Eh:-LEUVrR_3tsGIdIzzcc9:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/0f03fd127f534000deb09397eac53e5a15ed4113/email/templates/as_failure.html#L7)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

SCIONLab administrators
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224)*
<!-- Sent from Reviewable.io -->

mlegner commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 6 files at r1, 5 of 6 files at r2, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224#-:-LEUY1IfEeEvxnt6duSs:bnfp4nl)*
<!-- Sent from Reviewable.io -->

juagargi commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224)*
<!-- Sent from Reviewable.io -->

mlegner commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-coord/224)*
<!-- Sent from Reviewable.io -->
